### DISCUSSION TITLE
Use current buffer instead of pwd variable

### DISCUSSION THREAD

**Discussion by lagachea** [NeogitOrg/neogit#213] [Ideas]
Hi,

As of now the default behaviour when executing the `Neogit` function is the following:
- use the pwd variable to determine the repository. (at least from what I tested)

would it be possible to be:
- use the current buffer path to determine the repository.

I created a workaround in my config which has the desired effect, I `cd %:p:h` before running the `Neogit` function, the problem is that it changes the pwd variable when opening neogit.

PS: I am not an native english, excuse my poor grammar

**Comment by TimUntersberger** [comment]
What is your usecase for this?

> PS: I am not an native english, excuse my poor grammar

No worries.

**Reply by lagachea** [reply]
When working on repositoryA (regular work), I find myself needing to change some settings in my nvim config in repositoryB
For this I have a mapping using telescope to open the desired buffer in repoB.
Then to commit the changes made in repoB, I need to change the pwd variable using `cd %:p:h` to open Neogit inside repoB.
If I don't change the pwd variable and call `Neogit`, it opens in repoA.

**Reply by RianFuro** [reply]
Potentially also interesting for projects including submodules?
I'm a bit hesitant though, as `lcd` is a thing and usually solves this kind of multi-project scenarios, if done correctly.

@lagachea check out `:h lcd`. There's also `:h autochdir` if you prefer that behavior by default, but afaik that breaks committing through neogit (or has that been fixed while I wasn't paying attention @TimUntersberger?)

**Reply by TimUntersberger** [reply]
> or has that been fixed while I wasn't paying attention @TimUntersberger?

This should be fixed now.

> When working on repositoryA (regular work), I find myself needing to change some settings in my nvim config in repositoryB
For this I have a mapping using telescope to open the desired buffer in repoB.

This sounds like you should look into [sessions](https://vimhelp.org/starting.txt.html#Session)

**Comment by lagachea** [comment]
Thank you very much, I played with `lcd` and it is enough for my usecase.

Keep up the good work <3

**Reply by TimUntersberger** [reply]
> Thank you very much, I played with lcd and it is enough for my usecase.

That is good to here!

**Comment by pcoves** [comment]
Hi there.

I work with multiple layers of sub-modules and neither `autochdir` nor `lcd` are a good solutions for my use case.
I thought I could play with `neogit.open({cwd = foo})` but did not succeed here.

When using `magit` with `emacs` or `fugitive` with `vim/neovim`, the current buffer's project is used, not the `pwd` project.
How can I achieve the same result with `neogit`?

Many thanks in advance!
> P.

**Reply by pcoves** [reply]
Oopsy, a coworker of mine uses:

```lua
            vim.keymap.set("n", "<Leader>G", function()
                require("neogit").open({ cwd = vim.fn.expand('%:p:h') })
            end)
```

Seems to do just what I need.

**Reply by menash134** [reply]
Hi I also have a lot of git repositories in a project. I used the above keymap setting. It worked. But when I want to see the log message "?ll", then I get the error as below. Do you also have the same problem @pcoves. How do you solve it?
![image](https://github.com/NeogitOrg/neogit/assets/16444146/e635bac8-130a-444e-b7db-c1650825dc35)

**Reply by lkubb** [reply]
I had a similar issue while trying to open `NeogitLogCurrent` for the current buffer, regardless of the repo it's located in. I tried several things, but finally arrived at the following hacky solution (adjusted to work independent of my utility functions):

```lua
---Ensure that neogit popup actions open for the correct repo (the one that contains the current buffer).
---The only way to specify the cwd externally is to use neogit.open({cwd = '...'}), all other invocations (neogit.action)
---use the most recently requested repo instance.
---@param cmd string|table|fun() The command to execute. vim.cmd args or lua callback
local function neogit_cwd(cmd)
  local execute = function()
    if type(cmd) == 'function' then
      cmd()
    else
      vim.cmd(cmd)
    end
  end

  local buf_root = vim.fs.root(0, '.git')
  if not root then
    return execute()
  end

  local repo = require 'neogit.lib.git.repository'
  local neogit_root = repo.instance().worktree_root
  if neogit_root ~= buf_root then
    git.instance(buf_root):dispatch_refresh({ source = 'neogit_cwd', callback = vim.schedule_wrap(execute) })
  else
    execute()
  end
end
```

Can be used like this:

```lua
vim.keymap.set('n', '<leader>gH', function() ensure_cwd('NeogitLogCurrent') end)
vim.keymap.set('v', '<leader>gH', function()
  vim.cmd([[execute "normal! \<ESC>"]])
  local line_start = vim.fn.getpos("'<")[2]
  local line_end = vim.fn.getpos("'>")[2]
  ensure_cwd({ cmd = 'NeogitLogCurrent', range = { line_start, line_end } })
end)
```
